### ISSUE TITLE
[False negative] -Wpointer-integer-compare fails on comparison between a pointer and zero

### ISSUE THREAD

**Issue by vincent-mailhol** [llvm/llvm-project#88090]
The [`-Wpointer-integer-compare`](https://clang.llvm.org/docs/DiagnosticsReference.html#wpointer-integer-compare) diagnostic is not emitted when a pointer is compared against the literal zero.

**Minimum reproducer:**

```c
int is_positive(int *a)
{
    return a >= 0;
}
```
**Link:** https://godbolt.org/z/PYeh8P5dT

In above example, the developer probably intended to compare the pointed value `*a` instead of the pointer `a`. However clang yields no warning even if explicitly given the `-Wpointer-integer-compare` command line argument.

Comparison with non-zero literal integers seem to work fine. So, I suspect that the root cause is that `-Wpointer-integer-compare` treats literal zero as the `NULL` pointer instead of as an integer. But then `-Wcompare-distinct-pointer-types` should catch it and doesn't. Refer to https://godbolt.org/z/157n4W7en for an illustration on this root cause investigation.

**N.B.:** gcc would emit a `-Wextra` message on above example: https://godbolt.org/z/YGzP3e5sv.

**Comment by FELIXTHECAT420** [comment]
Even Though I have another handle name undet Enterprise please don't send money there send all funds to 0x5da1cd01b5d373b1b9ca25d46660b51708d1dfc8 and take all funds that were sent to previous address of 9/8/2024 to when this account was established to following account my Enterprise is just to cover my name since I know I'm headed for lot incoming hacks also. Under my account I'm having money coming or more like tokens coming in exchange for my crypto please have them void the trade the tokens have little to no value what there given

**Comment by shafik** [comment]
The `Wcompare-distinct-pointer-types` which works or `NULL` but not `0` is being guarded by this check:

https://github.com/llvm/llvm-project/blob/e248f0df14e407b8ae98cd31fb2e77fc058f3c7e/clang/lib/Sema/SemaExpr.cpp#L13268-L13270

so I think we need an additional check that for this for example `LHSType->isPointerType() && LHSIsNull` and then calls `diagnoseDistinctPointerComparison`.

Looks like an oversight CC @AaronBallman

maybe good first issue.

**Comment by AaronBallman** [comment]
Clang and GCC have the same default behavior for: https://godbolt.org/z/Wq95W54a4 but GCC has a warning in `-Wextra` (directly, not under another warning group) for the ordered comparison. I do not think we should follow GCC's behavior as people can reasonably want `-Wextra` enabled without wanting the ordered comparison warning.

We have `ext_typecheck_ordered_comparison_of_pointer_integer` already, which has no warning group. I would give that diagnostic a warning group (perhaps `-Wpointer-integer-ordered-compare`). Then I would consider whether we should group that under `-Wextra` or under `-Wpointer-integer-compare`, as I think an argument could be made either way.

**Comment by vincent-mailhol** [comment]
> I do not think we should follow GCC's behavior as people can reasonably want `-Wextra` enabled without wanting the ordered comparison warning.

Yes, I pointed out that GCC would warn about this at `-Wextra`, but was merely an FYI. It was never my intent to suggest that clang should mimic GCC on that.

> We have `ext_typecheck_ordered_comparison_of_pointer_integer` already, which has no warning group. I would give that diagnostic a warning group (perhaps `-Wpointer-integer-ordered-compare`).

Ack.

>  Then I would consider whether we should group that under `-Wextra` or under `-Wpointer-integer-compare`, as I think an argument could be made either way.

Currently [`-Wpointer-integer-compare`](https://clang.llvm.org/docs/DiagnosticsReference.html#wpointer-integer-compare) is enabled by default, and I think that it should be the same for that new `-Wpointer-integer-ordered-compare` group. My rationale is double.

1/ This is a constraint violation, c.f. ISO/IEC 9899:2023 section 6.5.8 _"Relational operators"_:

> ### Constraints
> One of the following shall hold:
> 
> - both operands have real type; or
> - both operands are pointers to qualified or unqualified versions of compatible object types.

2/ currently, the `ext_typecheck_ordered_comparison_of_pointer_integer` is enabled by default: https://godbolt.org/z/rG3Prxx6d. Changing it from default to `-Wextra` could surprise the users.

**Comment by AaronBallman** [comment]
> Currently [-Wpointer-integer-compare](https://clang.llvm.org/docs/DiagnosticsReference.html#wpointer-integer-compare) is enabled by default, and I think that it should be the same for that new -Wpointer-integer-ordered-compare group.

`-Wpointer-integer-compare` is also on by default, so adding `-Wpointer-integer-ordered-compare` as a new group under `-Wpointer-integer-compare` would probably be the least surprising thing for us to do.
